src/core/memory.cc

```diff
@@ -1926,14 +1926,14 @@ void* realloc(void* ptr, size_t size) {
         // If ptr is a null pointer, the behavior is the same as calling std::malloc(new_size).
         return malloc(size);
     } else if (!is_seastar_memory(ptr)) {
-        // we can't realloc foreign memory on a shard
-        if (is_reactor_thread) {
-            abort();
-        }
         // original_realloc_func might be null when previous ctor allocates
         if (original_realloc_func) {
             return original_realloc_func(ptr, size);
         }
+        // we can't realloc foreign memory on a shard before original_realloc_func is set
+        if (is_reactor_thread) {
+            abort();
```

Comment:
On second reading, I'm even more confused. If we are in the `!is_seastar_memory`, I think we always need to call original_realloc_func(), we can't fall back to the irelevant code below. So why shouldn't we call `abort()` unconditionally? Why if it's NOT a reactor thread (i.e., an alien thread?) is it fine to fall through to the irrelevant code below?

Reply:
See the comment below ("if we're here"). IIRC it is reached when some code around boost program_options is reached early.

Reply:
But you just told me, when I asked doesn't free() have the same problem, that Seastar's free() does work correctly with libc allocation...
Also, as as you said now, if the free() **doesn't** work, then the fallthrough is wrong and we should abort (?).
If free() **does** work, I don't understand why we need to abort at all.

Reply:
::free() works for non-seastar memory if original_free_func is set, but we don't know if it is.

The code below won't work for non-seastar memory, so I guess the fallthrough is wrong.